### PR TITLE
[TASK] Implement PostgreSQL tables for maps

### PR DESCRIPTION
**Pull request by zdigness**
## Description
The goal of this ticket is to:

Create a PostgreSQL table for maps, sandtraps, barriers, and water

Maps:
- associated with a day (YYYY-MM-DD)
- contains ball x and y positions
- contains hole x and y positions
- one-to-many with sandtraps, water, and barriers

Barriers, sandtraps, water:
- associated with a map ID
- contains x and y positions

## Motivation and Context
This allows us to load the puttdle game scene according to the current date

We now have to connect this with the FE and call api/v1/map route to get the data for the map of the day

## How Has This Been Tested?

1. Mocked using Jest
2. Tested using Postman and pgAdmin (Video below)

Quick queries for testing:
INSERT INTO "Maps" (day, par, hole_x, hole_y, ball_x, ball_y, "createdAt", "updatedAt")
VALUES ('2024-04-15', 72, 200, 150, 220, 175, NOW(), NOW());

INSERT INTO "Barriers" ("mapId", x, y, "createdAt", "updatedAt")
VALUES (6, 50, 100, NOW(), NOW());

## Artifacts (if appropriate):
(the first minute is the container starting mb)

https://github.com/zdigness/puttdle/assets/60762808/6399b733-fa5c-4236-881b-aec0ee7369aa

### CODE REVIEW

File: server/docker-compose.yml
Comment (zdigness) [inline server/docker-compose.yml:1]: Random auto formatting

File: server/src/controllers/__tests__/map.spec.ts
Comment (zdigness) [inline server/src/controllers/__tests__/map.spec.ts:1]: mock the map controller

File: server/src/controllers/map.ts
Comment (zdigness) [inline server/src/controllers/map.ts:1]: only task we need in the controller is to get the map for current date from the database

File: server/src/db/index.ts [hunk 1]
```diff
@@ -31,11 +35,15 @@ class Database {
...
-    models: [User, Score],
+    models: [User, Score, Map, Sandtrap, Water, Barrier],
   })
 
   public User = this.sequelize.getRepository(User)
   public Score = this.sequelize.getRepository(Score)
+  public Map = this.sequelize.getRepository(Map)
+  public Sandtrap = this.sequelize.getRepository(Sandtrap)
+  public Water = this.sequelize.getRepository(Water)
+  public Barrier = this.sequelize.getRepository(Barrier)
```
Comment (zdigness) [inline server/src/db/index.ts:46]: add new tables to the ORM

File: server/src/routes/map.ts
Comment (zdigness) [inline server/src/routes/map.ts:1]: http://localhost:3000/api/v1/map

File: server/src/controllers/__tests__/map.spec.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    const consoleSpy = jest.spyOn(console, 'error');
+    (database.Map.findOne as jest.Mock).mockImplementation(() => {
+      throw new Error('Test error');
+    });
+
+    const fullMap: FullMap | null = await MapController.getMap(new Date());
+
+    expect(fullMap).toBeNull();
+    expect(consoleSpy).toHaveBeenCalled();
+  });
```
Comment (AronPerez) [inline server/src/controllers/__tests__/map.spec.ts:78]: We should be `describing` `getMap`. Could you please put these tests behind a describe block for that?

File: server/docker-compose.yml [hunk 3]
```diff
@@ -16,7 +16,7 @@ services:
       - "3000:3000"
     depends_on:
       - db
-    command: ["npm", "run", "dev"]
+    command: [ "npm", "run", "dev" ]
```
Comment (AronPerez) [inline server/docker-compose.yml:19]: Why were spaces added?

File: server/src/models/barrier.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+  @ForeignKey(() => Map)
+  @Column
+    mapId!: number
+
+  @Column(DataType.INTEGER)
+    x!: number
+
+  @Column(DataType.INTEGER)
+    y!: number
```
Comment (AronPerez) [inline server/src/models/barrier.ts:17]: Do you think we should make these column names more descriptive like your naming convention in `map.ts` with `barrier_x` or maybe even `barrier_x_cord` (but we'd need to update other records using similar coordinates)?

File: server/src/routes/map.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+const mapRouter = Router()
+
+mapRouter.get("/", async (req, res) => {
+  try {
+    const today = new Date()
+    console.log(today)
+    const mapExists: FullMap | null = await MapController.getMap(today)
+
+    if (mapExists) {
+      const fullMap: FullMap | null = await MapController.getMap(today)
```
Comment (AronPerez) [inline server/src/routes/map.ts:14]: Why are we getting the map twice?

File: server/src/routes/map.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+const mapRouter = Router()
+
+mapRouter.get("/", async (req, res) => {
+  try {
+    const today = new Date()
+    console.log(today)
+    const mapExists: FullMap | null = await MapController.getMap(today)
+
+    if (mapExists) {
```
Comment (AronPerez) [inline server/src/routes/map.ts:13]: So what's happening if the map doesn't exist for a day? Shouldn't we return the status `404`?

File: server/src/models/water.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+  @ForeignKey(() => Map)
+  @Column
+    mapId!: number
+
+  @Column(DataType.INTEGER)
+    x!: number
+
+  @Column(DataType.INTEGER)
+    y!: number
```
Comment (AronPerez) [inline server/src/models/water.ts:17]: See note above

File: server/src/models/sandtrap.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+  @ForeignKey(() => Map)
+  @Column
+    mapId!: number
+
+  @Column(DataType.INTEGER)
+    x!: number
+
+  @Column(DataType.INTEGER)
+    y!: number
```
Comment (AronPerez) [inline server/src/models/sandtrap.ts:17]: See note above

File: server/src/models/map.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    ball_y!: number
+
+  @HasMany(() => Sandtrap)
+    sandtraps!: Sandtrap[]
+
+  @HasMany(() => Water)
+    waters!: Water[]
+
+  @HasMany(() => Barrier)
+    barriers!: Barrier[]
```
Comment (AronPerez) [inline server/src/models/map.ts:45]: Good job setting this up, nice and clean 🧼

File: server/src/controllers/__tests__/map.spec.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
+import MapController from '../map';
+import database from '../../db/index';
+import { Repository, Model } from 'sequelize-typescript';
+import { Map } from '../../models/map';
+import { Sandtrap } from '../../models/sandtrap';
+import { Water } from '../../models/water';
+import { Barrier } from '../../models/barrier';
+import { FullMap } from '../../types';
```
Comment (AronPerez) [inline server/src/controllers/__tests__/map.spec.ts:8]: Can you run the linter on this file, please? It should have caught this in the linter action; I will need to update that.

**Review by AronPerez — CHANGES_REQUESTED** [review]
When you can, please look at the following:
1. Naming convention for columns in regards to coordinates
2. Map route logic
3. Map route resolution scenarios
4. Map controller spec test

File: server/docker-compose.yml (hunk 3 above)
Comment (zdigness) [reply server/docker-compose.yml:19]: It kept auto formatting for me whenever I tried to change it back and save

File: server/src/routes/map.ts (hunk 6 above)
Comment (haydenbalsys) [reply server/src/routes/map.ts:13]: That solution should work for now

**Review by haydenbalsys — COMMENTED** [review]
Looks good. What is a controller in this case? e.g. Why do we have a controllers directory with the tests in it?

**Comment by AronPerez** [comment]
@haydenbalsys It's not, @zdigness needs to update the following:
1. Naming convention for columns in regards to coordinates
2. Map route logic
3. Map route resolution scenarios

@haydenbalsys We're using MVC architecture, I'd [read this](https://blog.logrocket.com/building-structuring-node-js-mvc-application/) or [watch this](https://www.youtube.com/watch?v=zW_tZR0Ir3Q) if you're curious. 

As for having a `__tests__` directory in corresponding directories, we can either have one directory in the root of `server/` or what we're doing. I've chosen this approach since it isolates tests to their respective source files and is easier to manager. Check [here](https://jestjs.io/docs/more-resources#learn-by-example) for more examples.

File: server/src/routes/map.ts (hunk 6 above)
Comment (AronPerez) [reply server/src/routes/map.ts:13]: We need a 404 if we request a map that doesn't exist. There isn't a reason we shouldn't be handling it

File: server/.prettierrc [hunk 11]
```diff
@@ -1,5 +1,4 @@
 {
-	"endOfLine": "lf",
```
Comment (AronPerez) [inline server/.prettierrc:2]: Removed since Github is setting to `CLRF`

File: server/src/controllers/__tests__/map.spec.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+      expect(database.Sandtrap.findAll).toHaveBeenCalledWith({ where: { mapId: 1 } })
+      expect(database.Water.findAll).toHaveBeenCalledWith({ where: { mapId: 1 } })
+      expect(database.Barrier.findAll).toHaveBeenCalledWith({ where: { mapId: 1 } })
+      expect(fullMap).toEqual(expectedFullMap)
+    })
+
+    it("should return null when the map does not exist", async () => {
+      ;(database.Map.findOne as jest.Mock).mockResolvedValue(null)
+
+      const fullMap: FullMap | null = await MapController.getMap(new Date())
```
Comment (AronPerez) [inline server/src/controllers/__tests__/map.spec.ts:64]: What happens if we have multiple maps with the same date?

File: server/src/models/sandtrap.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+  @ForeignKey(() => Map)
+  @Column
+  mapId!: number
+
+  @Column(DataType.INTEGER)
+  sandtrap_x!: number
+
+  @Column(DataType.INTEGER)
+  sandtrap_y!: number
```
Comment (AronPerez) [inline server/src/models/sandtrap.ts:17]: Well done

File: server/src/models/map.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+@Table({ tableName: "Maps" })
+export class Map extends Model<Map> {
+  @PrimaryKey
+  @AutoIncrement
+  @Column(DataType.INTEGER)
+  id!: number
+
+  @AllowNull(false)
+  @Column(DataType.DATEONLY)
+  day!: Date
```
Comment (AronPerez) [inline server/src/models/map.ts:16]: Since we're consistently doing lookups on the date column, indexing it for quicker lookups would be ideal.

**Review by AronPerez — DISMISSED** [review]

**Review by AronPerez — APPROVED** [review]

**Pull request merged**